### PR TITLE
move dt_optimize() into DtBuilder.finish()

### PR DESCRIPTION
**Pull request by WalterBright**
Also add overload to DtBuilder.cat() to avoid more pointer chasing.

### CODE REVIEW

File: src/backend/out.c [hunk 1]
```diff
@@ -159,7 +159,8 @@ void outdata(symbol *s)
                 //printf("DT_azeros %d\n", dt->DTazeros);
             case_azeros:
                 datasize += dt->DTazeros;
-                if (dt == dtstart && !dt->DTnext && s->Sclass != SCcomdat)
+                if (dt == dtstart && !dt->DTnext && s->Sclass != SCcomdat &&
+                    (s->Sseg == UNKNOWN || s->Sseg <= UDATA))
```
Comment (9rnsr) [inline src/backend/out.c]: How this is related with?

**Comment by 9rnsr** [comment]
Other than that, looks good to me.

**Comment by braddr** [comment]
Hrm.. the two optimizations are different.  The existing code collapses runs anywhere in the list.  The new code only at the head of the list.  Why not just collapse while adding in the first place and bypass the extra memory allocation and need to re-traverse to collapse?

File: src/backend/out.c (hunk 1 above)
Comment (WalterBright) [inline src/backend/out.c]: This code puts DTazeros into the BSS segment. There is code in objc_glue.c that was assigning a different seg number, and expecting the data to be put into that segment. Expanding the azeros optimizations broke that. This fixes it.

**Comment by WalterBright** [comment]
> Why not just collapse while adding in the first place and bypass the extra memory allocation and need to re-traverse to collapse?

Why not indeed. The main reason for collapsing at the front is so that "allzeros" can do its thang. Collapsing in the middle doesn't accomplish much of anything.

As Martin suggested, and what all these changes enable, is changing the data structure. It should probably be an array, not a linked list.

**Comment by 9rnsr** [comment]
Auto-merge toggled on

**Pull request merged**